### PR TITLE
Add a super-grouped grid layout transformation

### PR DESCRIPTION
**Pull request by giuseros**
This PR is aiming at imposing an layout on the distribution of the workgroups, aiming at minimize cache trashing. The idea is the same of: https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html#sphx-glr-getting-started-tutorials-03-matrix-multiplication-py

The main changes are:
- I added a transformation to map `bid` -> `g_block, m_block, n_block` to mimic what Triton does. Since we cannot have a piecewise affine transformation, I had to manually insert the transformation directly in MLIR. This is not super elegant and might be corrected in followup PR. 
- The group size is calculated via a simple heuristic on the number of CUs (`groupSize=sqrt(numCU)`). For now we don't add any additional tuning parameter
- `numCU` is now mandatory in gridwise gemm and it is used as part of the problem description (i.e., `numCU` will be part of the key in the tuning db). This means I had to change `perfRunner.py` and `tuningRunner.py`.

### CODE REVIEW

File: mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp]: Can we not re-calculate grid coordinates here from gridSize ? 
The caller of computeOutputTransforms should already have bidGridLengths.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: We dont need to this after https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/1134.
i.e. we dont need this logic recalculated at this level in the lowering.

Please have a look at the PR and I think could simply pass in Values : g_block, m_block and n_block when threadwise_read / threadwise_write ops are created.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: Same here... (see the above comment)

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:380]: This could all just be a threadwise_read_into... 
i.e. no need to repeat the logic

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: This is not right...

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: As this is not a transform pass, IMO this file should live in utility? 
(same could be said for AccelEmitter)

**Review by manupak — DISMISSED** [review]
The intent and outcome is sound! I agree with the direction.

I just have concerns with layering.
I am doing some cleanup/improvement on parallel to increase modularity/extensibility of the codebase that could be useful here. 

Please have a look at the PR : 
1) https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/1134

With the above, you shouldn't need to redo the logic restore block id in the lowering of threadwise_read_into / write_all ops. I believe you already have them when those ops are created .. therefore, you just pass them as "extraIndices" field.

2) AccelEmitter should not be re-calculating grid layout back from gridSize. Im also doing something here : https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/1138. One of the changes here is to pass the bidGridLengths to accelEmitter. I believe with that you dont need to re-do g_block, m_block and n_block extractions again in AccelEmitter.

So generally, my comments are to push to a singular place where this change should happen and others should 'just' work(TM) :) .

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: I think Transforms/ isn't required to just be passes, it can also have pass-specific support code

Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: which I always thought why Transforms/Utils/ or Utils/ exist... (but yea it is not strictly followed upstream either)
Anyways I dont have a strong opinion; just felt nice to have an criterion to decide which goes where.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:362]: Instead of bid here,  lets pass in the g_block val, m_block val and n_block val if this function is being kept.
I still think we can replace the whole function with a threadwise_read_jnto.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: you'll need to pass in extraIndices here as  = {g_block, m_block, n_block, tid} to match the top dims of idToMatrixCMaps

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td
Comment (manupak) [inline mlir/include/mlir/Dialect/Rock/IR/RockOps.td:118]: why not OptionalAttr on numCU here ?

**Review by manupak — COMMENTED** [review]
few further pointers/comments...

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: Yeah I don't know, I mean my (also not strong) preference is more toward a flat structure. I guess if we reach a point where we have a lot of utility code, then we can move everything in a separate folder

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td
Comment (giuseros) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td:118]: This was like this because the `numCU` is mandatory in backward weight convolution (because it is needed to compute the `KBlock` and the workspace)

**Comment by giuseros** [comment]
Hi @manupak , I rebased and restructured the code like you proposed, and I agree it's way cleaner. Thanks!

File: mlir/include/mlir/Dialect/Rock/IR/RockGemmWrapperInterface.td
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/IR/RockGemmWrapperInterface.td]: Failed to update comment

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: What's this include for?

File: mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/utility/AmdArchDb.cpp]: Can we get the little informative comments uniformly?

File: mlir/utils/performance/perfRunner.py
Comment (krzysz00) [inline mlir/utils/performance/perfRunner.py]: Stray debug print

**Review by krzysz00 — COMMENTED** [review]
Minor nitpicks

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: nit (naming) : this could be GridLayoutEmitter as it not a generic layout emitter

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.cpp]: nit (naming) : "rock-grid-layout-emitter" using this in CLI with -debug-only="layout-transform" might be a little too vague.

File: mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.h
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/LayoutEmitter.h]: nit (naming) : this could be "makeGroupedGridLayout" just because we might end up different ways here

**Review by manupak — COMMENTED** [review]
Looking good! few naming nit picks.

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td
Comment (krzysz00) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td:118]: It's only needed for atomic_add kernels, though, so it's kinda optional still. On the other hand, MIOpen integration is off so _shrug_.

**Review by krzysz00 — COMMENTED** [review]
The code here seems fine to me.

Do we have perf numbers?

**Comment by giuseros** [comment]
> The code here seems fine to me.
> 
> Do we have perf numbers?

Yes, they are coming.  Need to collect base results and then we can compare

**Review by manupak — COMMENTED** [review]
One doc fix that you prolly missed otherwise looks good on the condition perf agree to numCU based (full occupancy) heuristic. 

I ll wait for perf before hitting approve.

File: mlir/lib/Dialect/Rock/Transforms/GridLayoutEmitter.h
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridLayoutEmitter.h]: I think you missed this.

**Comment by giuseros** [comment]
As I mentioned in the meeting, I have had an issue with `i8xi8->i32` performance. 

Basically, the fact that threads were loading i8 and storing i32 were messing up with the cache (I think). I added a sort of normalization factor based on the ratio `outBitwidth/inBitwidth` that sorted out the issue. I am attaching here the comparison:

[perf.zip](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/12074095/perf.zip)

**Comment by giuseros** [comment]
A friendly reminder @manupak  @krzysz00 :)

File: mlir/lib/Dialect/Rock/Transforms/GridLayoutEmitter.cpp
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridLayoutEmitter.cpp:52]: Im just realizing we do 1-d tiling here... why ?

Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridLayoutEmitter.cpp:52]: and this is not equivalent to 

```py
# Number of programs ids along the N axis
num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
# Number of programs in group
num_pid_in_group = GROUP_SIZE_M * num_pid_n
```

where num_pid_n is value of current n_block rather than n_blocks right ?

**Review by manupak — COMMENTED** [review]
Sorry (didnt catch this early), I have last min question about the logic...

File: mlir/lib/Dialect/Rock/Transforms/GridLayoutEmitter.cpp
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridLayoutEmitter.cpp:52]: oops, yes it is

**Comment by manupak** [comment]
(it was my bad)

**Review by manupak — APPROVED** [review]
Codewise approved!
perf-wise I can see -6.94 to +10 swing.. 

do you have any idea for the negative ones ?

**Comment by giuseros** [comment]
My guess would be that the heuristic we are using is too simple. I think this PR will enable us to do this sort of investigation, hoping not only to sort out the regressions you mention, but also to improve performance further.

**Comment by giuseros** [comment]
I created an investigation ticket here: https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/1022

**Pull request merged**
